### PR TITLE
GODRIVER-2222 Prevent removing valid servers when updating SRV hosts.

### PR DESCRIPTION
**Pull request by matthewdale**
[GODRIVER-2222](https://jira.mongodb.org/browse/GODRIVER-2222)

Currently, the driver may remove valid servers from a topology when updating the topology based on the SRV host records if `len(SRV hosts) > srvMaxHosts`. Instead, when updating topology servers from SRV host records, wait until we've determined the added and removed hosts to limit the topology servers list to `srvMaxHosts`.

Also use `rand.Rand.Shuffle` instead of a custom shuffle implementation and re-enable the "SRVMaxHosts is less than number of hosts" test.

### CODE REVIEW

File: x/mongo/driver/topology/topology.go [hunk 1]
```diff
@@ -597,11 +584,22 @@ func (t *Topology) processSRVResults(parsedHosts []string) bool {
 		t.fsm.removeServerByAddr(addr)
 		t.publishServerClosedEvent(s.address)
 	}
+
+	// Now that we've removed all the hosts that disappeared from the SRV record, we need to add any
+	// new hosts added to the SRV record. Shuffle the list of added hosts so we add them in random
+	// order. Stop adding hosts whenever we reach srvMaxHosts.
+	random.Shuffle(len(diff.Added), func(i, j int) {
```
Comment (benjirewis) [inline x/mongo/driver/topology/topology.go:591]: Shouldn't we only `Shuffle` if `srvMaxHosts > 0`?

**Review by benjirewis — COMMENTED** [review]
Thanks for taking this!

File: x/mongo/driver/topology/topology.go (hunk 1 above)
Comment (kevinAlbs) [reply x/mongo/driver/topology/topology.go:591]: The [Polling SRV records specification](https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst#querying-dns) indicates shuffling only if necessary:

> If srvMaxHosts is zero or greater than or equal to the number of hosts in the DNS result, the driver MUST populate the seedlist with all hosts.
>
> If srvMaxHosts is greater than zero and less than the number of hosts in the DNS result, the driver MUST randomly select that many hosts and use them to populate the seedlist. Drivers SHOULD use the Fisher-Yates shuffle for randomization.

A possible impact of always shuffling is a shuffled order of ServerOpening events.

I agree with the suggested change of checking if `srvMaxHosts == 0` to shuffle along with the condition `srvMaxHosts < len(diff.Added)`.

**Review by kevinAlbs — APPROVED** [review]
LGTM with suggested change to only shuffle when necessary.

File: x/mongo/driver/topology/topology.go (hunk 1 above)
Comment (matthewdale) [reply x/mongo/driver/topology/topology.go:591]: Good idea, will update.

**Review by benjirewis — APPROVED** [review]
LGTM!

**Comment by matthewdale** [comment]
All failing tests are due to [DRIVERS-1991](https://jira.mongodb.org/browse/DRIVERS-1991) or [GODRIVER-2238](https://jira.mongodb.org/browse/GODRIVER-2238).

**Pull request merged**